store/postgres/src/store.rs

```diff
@@ -765,6 +668,128 @@ impl StoreTrait for Store {
+        node_id: NodeId,
+    ) -> Result<(), Error> {
+        use db_schema::subgraph_deployments;
+
+        insert_into(subgraph_deployments::table)
+            .values((
+                subgraph_deployments::deployment_name.eq(name.to_string()),
+                subgraph_deployments::subgraph_id.eq(subgraph_id.to_string()),
+                subgraph_deployments::node_id.eq(node_id.to_string()),
+            )).on_conflict(subgraph_deployments::deployment_name)
+            .do_update()
+            .set((
+                subgraph_deployments::subgraph_id.eq(subgraph_id.to_string()),
+                subgraph_deployments::node_id.eq(node_id.to_string()),
+            )).execute(&*self.conn.lock().unwrap())
+            .map_err(Error::from)
+            .map(|_| ())
+    }
+
+    fn read(&self, name: SubgraphDeploymentName) -> Result<Option<(SubgraphId, NodeId)>, Error> {
+        use db_schema::subgraph_deployments;
+
+        subgraph_deployments::table
+            .select((
+                subgraph_deployments::subgraph_id,
+                subgraph_deployments::node_id,
+            )).filter(subgraph_deployments::deployment_name.eq(name.to_string()))
+            .first::<(String, String)>(&*self.conn.lock().unwrap())
+            .optional()
+            .map_err(Error::from)
+            .map(|row_opt| {
+                row_opt.map(|(subgraph_id, node_id)| {
+                    let subgraph_id = SubgraphId::new(subgraph_id)
+                        .expect("invalid subgraph ID found in database");
+                    let node_id = NodeId::new(node_id).expect("invalid node ID found in database");
+                    (subgraph_id, node_id)
+                })
+            })
+    }
+
+    fn remove(&self, name: SubgraphDeploymentName) -> Result<bool, Error> {
+        use db_schema::subgraph_deployments;
+
+        delete(subgraph_deployments::table)
+            .filter(subgraph_deployments::deployment_name.eq(name.to_string()))
+            .execute(&*self.conn.lock().unwrap())
+            .map(|row_count| match row_count {
+                0 => false,
+                1 => true,
+                _ => unreachable!(),
+            }).map_err(Error::from)
+    }
+
+    fn deployment_events(
+        &self,
+        node_id: NodeId,
+    ) -> Box<Stream<Item = DeploymentEvent, Error = Error> + Send> {
+        // Create postgres listener
+        let channel_name =
+            SafeChannelName::i_promise_this_is_safe(format!("subgraph_deployment_{}", node_id));
+        let mut listener = NotificationListener::new(self.postgres_url.clone(), channel_name);
+
+        // Start receiving notifications
+        listener.start();
+
+        Box::new(
+            listener
+                .take_event_stream()
+                .unwrap()
+                .map(move |notification| {
+                    // Move listener into closure so that listener lives as long as stream does
+                    let _ = listener;
```

Comment:
I always thought that this would drop the value! Definitely learned something here.

Reply:
Looking at this again, I'm actually surprised it compiles... it looks like to me the closure should be `FnOnce`, because `listener` is being moved into a variable. But I guess it's either a special case on `_`, or the fact that the variable is unused, that means that `listener` can be taken by reference for the right-hand expression? Magic! :sparkles:

Reply:
Yhea apparently this is defined as a noop. Funny that you did the magic trick without really thinking about it.